src/memtester_parsing.py

```diff
@@ -0,0 +1,189 @@
+        
+    @_(r"0x[0-9a-f]+")
+    def HEX(self, t):
+        t.value = int(t.value, 16)
+        return t
+    
+    FLR = r"FAILURE"
+    OK = r"ok"
+    
+    # String containing ignored characters (between tokens)
+    ignore = " \n\t\b:\\/|!=.-"
+    
+    # Ignored sequences
+    ignore_header = r"(Copyright|Licensed|pagesize|pagesizemask|want|got).+\n"
+    ignore_status = r"(testing|setting)\s+[0-9]+"
+    ignore_flr1_desc = r"possible\sbad\saddress\sline\sat\s(physical\saddress|offset)"
+    ignore_flr3_desc = r"at\s(physical\saddress|offset)"
+    ignore_skip = r"Skipping\sto\snext\stest\.\.\."
+    ignore_done = r"Done\."
+    
+    def __init__(self, callbacks: MemtesterCallbacks):
+        self._callbacks = callbacks
+    
+    def error(self, t):
+        m = "Illegal character '{}' at line {}:{}".format(t.value[0], t.lineno, t.index)
+        self._callbacks.parsing_error(m);
+        self.index += 1
+
+
+# https://sly.readthedocs.io/en/latest/sly.html#writing-a-parser
+class MemtesterParser(Parser):
+    # A list of supported memtester versions
+    _known_versions = ["4.6.0", "4.5.1", "4.5.0"]
+
+    # Get the token list from the lexer (required)
+    tokens = MemtesterLexer.tokens
+
+    def __init__(self, callbacks: MemtesterCallbacks):
+        self._callbacks = callbacks
+
+    @_("version loops")  
+    def run(self, p):
+        self._callbacks.run_ready(p.loops)
+
+    @_("VER")  
+    def version(self, p):
+        self._callbacks.version_ready(p.VER, p.VER in self._known_versions)
+        
+    @_("loops loop", "loop")   
+    def loops(self, p):
+        self._callbacks.loop_ready(p.loop)
+        
+        # Count failed loops
+        prev_failed_loops = p.loops if len(p) == 2 else 0
+        curr_failed_loop = 1 if len(p.loop.failed_tests()) > 0 else 0
+        return prev_failed_loops + curr_failed_loop
+        
+    @_("LOOP tests")   
+    def loop(self, p):
+        return Loop(p.LOOP, p.tests)
+    
+    @_("tests test", "test")     
+    def tests(self, p):
+        self._callbacks.test_ready(p.test)
+        return ([] if len(p) == 1 else p.tests) + [p.test]
+    
+    @_("TEST OK")     
+    def test(self, p):
+        return Test(p.TEST, OkResult())
+        
+    @_("TEST failures")     
+    def test(self, p):
+        return Test(p.TEST, p.failures)
+
+    @_("failures failure", "failure")
+    def failures(self, p):
+        return ([] if len(p) == 1 else p.failures) + [p.failure]
+        
+    @_("FLR HEX HEX HEX")
+    def failure(self, p):
+        return BadAddress(addr=p.HEX2, left_val=p.HEX0, right_val=p.HEX1)      
+        
+    @_("FLR HEX")
+    def failure(self, p):
+        return BadLine(addr=p.HEX) 
+    
+    def error(self, p):
+        if p: # Do not report EOF as error
+            m = "{} was unexpected".format(p)
+            self._callbacks.parsing_error(m)
+
+      
+class MemtesterObserver:
+    def __init__(self):
+        self.callbacks = MemtesterCallbacks()
```

Comment:
all fields of `self.callbacks` are None here; so if this `MemtesterObserver` were to be used as it's constructed here, any caller would fail while trying to call `None` callables. As such, this object instantiation here is incomplete

we should have "correct by construction` semantics for all of the objects in this codebase. this correlates with my prev comment on the callbacks dataclass. if that was a normal class with default impls for the callback methods, this instance here would be correct

Reply:
`MemtesterCallbacks` initializes all its members with [empty lambdas](https://github.com/opencomputeproject/ocp-diag-memtester/blob/7d11a44daad738e603a8b27b201b5be79843482a/src/memtester_parsing.py#L45).

Reply:
sorry, i missed the empty lambdas in the initial review; saw `None` and shortcircuited to `x = None` instead of `x = lambda: None`

my other comment about OOP still stands